### PR TITLE
KAFKA-16248

### PR DESCRIPTION
**Pull request by aliehsaeedii**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java [hunk 1]
```diff
@@ -90,6 +91,10 @@ public int count() {
         return count;
     }
 
+    public Map<TopicPartition, OffsetAndMetadata> nextOffsets() {
+        return new HashMap<>();
+    }
+
```
Comment (aliehsaeedii) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRecords.java:97]: will be implemented.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 2]
```diff
@@ -188,6 +190,7 @@ public StreamTask(final TaskId id,
 
         // initialize the consumed and committed offset cache
         consumedOffsets = new HashMap<>();
+        consumedLeaderEpochs = new HashMap<>();
```
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: This collection seems to be only written, not read. Do we need it?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 3]
```diff
@@ -278,6 +281,10 @@ public void initializeIfNeeded() {
         }
     }
 
+    public void setConsumedOffsetsAndMetadata(final Map<TopicPartition, OffsetAndMetadata> consumedOffsetsAndMetadata) {
+        this.consumedOffsetsAndMetadata = consumedOffsetsAndMetadata;
```
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: I think these are not the already consumed offsets and metadata but the next offsets to be consumed, right? I would reflect this in the naming.

Maybe `nextOffsetsAndMetadataToBeConsumed`, even if lengthy.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java [hunk 4]
```diff
@@ -96,6 +96,7 @@ private int processTask(final Task task, final int maxNumRecords, final long beg
         try {
             while (processed < maxNumRecords && task.process(now)) {
                 task.clearTaskTimeout();
+                task.setConsumedOffsetsAndMetadata(taskManager.offsetAndMetadataPerTask().get(task));
```
Comment (lucasbru) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: It seems we are updating the consumedoffsetsAndMetadata in every single iteration of the process loop. But it seems, they change only once after every poll.

It seems we are already passing the `ConsumerRecords` object into `addRecordsToTasks`. Couldn't we call `task.setConsumedOffsetsAndMetadata` once inside `addRecordsToTasks` (for all tasks)?

**Review by lucasbru — APPROVED** [review]
Hi @aliehsaeedii ! I think the logic is correct. I wonder if we can simplify it slightly. I left some comments

**Review by lucasbru — COMMENTED** [review]
Hi @aliehsaeedii ! I think the logic is correct. I wonder if we can simplify it slightly. I left some comments

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java (hunk 4 above)
Comment (aliehsaeedii) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: good point but adding records to tasks does not mean that they have been processed.

**Pull request closed without merging**